csrc/welford.cu

```diff
@@ -0,0 +1,429 @@
+    grad_weight[blockIdx.x] = static_cast<scalar_t>(sum_dy_xmu_l[0] * factor);
+    mean_dy[blockIdx.x] = sum_dy_l[0] / total_item_num;
+    mean_dy_xmu[blockIdx.x] = sum_dy_xmu_l[0] / total_item_num;
+  }
+}
+
+template <typename scalar_t, typename accscalar_t>
+__global__ void batchnorm_backward_kernel(
+      const scalar_t* __restrict__ grad_output,
+      const scalar_t* __restrict__ input,
+      const scalar_t* __restrict__ mean,
+      const scalar_t* __restrict__ var,
+      const scalar_t* __restrict__ weight,
+      const scalar_t* __restrict__ mean_dy,
+      const scalar_t* __restrict__ mean_dy_xmu,
+      scalar_t* __restrict__ grad_input,
+      const int ss,
+      const float eps) {
+  int address_base = blockIdx.x*ss + blockIdx.y*gridDim.x*ss;
+
+  auto m_c = static_cast<accscalar_t>(mean[blockIdx.x]);
+  auto m_dy_c = static_cast<accscalar_t>(mean_dy[blockIdx.x]);
+  auto factor_1_c = static_cast<accscalar_t>(var[blockIdx.x]) + eps;
+  auto factor_2_c = static_cast<accscalar_t>(weight[blockIdx.x]) / sqrt(factor_1_c);
+  factor_1_c /= static_cast<accscalar_t>(mean_dy_xmu[blockIdx.x]);
+
+  // sequential welford
+  for (int offset = threadIdx.x; offset < ss ; offset+= blockDim.x) {
+    grad_input[address_base+offset] = (static_cast<accscalar_t>(grad_output[address_base+offset]) - m_dy_c - (static_cast<accscalar_t>(input[address_base+offset]) - m_c) / factor_1_c) * factor_2_c;
+  }
+}
+
+std::vector<at::Tensor> welford_mean_var_CUDA(const at::Tensor input) {
+  const auto batch_size = input.size(0);
+  const auto feature_size = input.size(1);
+
+  auto space_size = input.size(2);
+  for (int i = 3; i < input.ndimension(); i++) {
+    space_size *= input.size(i);
+  }
+
+  at::Tensor out_var = at::empty_like(input).resize_({feature_size});
+  at::Tensor out_var_biased = at::empty_like(out_var);
+  at::Tensor out_mean = at::empty_like(out_var);
+
+  int block_x = 16;
+  const dim3 block(block_x, batch_size);
+  const dim3 grid(feature_size);
+  // save current mean, var, num_elements;
+  int smem_size = batch_size * block_x * 3 * sizeof(int);
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.type(), "welford_mean_var_kernel", ([&] {
+    using accscalar_t = at::acc_type<scalar_t, true>;
+    welford_kernel<scalar_t, accscalar_t><<<grid, block, smem_size>>>(
+        input.data<scalar_t>(),
+        out_mean.data<scalar_t>(),
+        out_var.data<scalar_t>(),
+        out_var_biased.data<scalar_t>(),
+        batch_size,
+        feature_size,
+        space_size);
+  }));
+
+  return {out_mean, out_var, out_var_biased};
+}
+
+at::Tensor batchnorm_forward_CUDA(
+    const at::Tensor input,
+    const at::Tensor mean,
+    const at::Tensor var,
+    const at::Tensor weight,
+    const at::Tensor shift,
+    const float eps) {
+  const auto batch_size = input.size(0);
+  const auto feature_size = input.size(1);
+  at::Tensor out = at::empty_like(input);
+
+  auto space_size = input.size(2);
```

Comment:
This setup code is in every function, should be refactored into a separate function